crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,544 @@
+            .collect()
+    };
+    circuit.push_packed_operation(
+        PackedOperation::from_standard_instruction(StandardInstruction::Barrier(num_qubits)),
+        &[],
+        &qubits,
+        &[],
+    );
+    ExitCode::Success
+}
+
+#[repr(C)]
+pub struct OpCount {
+    name: *mut c_char,
+    count: usize,
+}
+
+#[repr(C)]
+pub struct OpCounts {
+    data: *mut OpCount,
+    len: usize,
+}
+
+/// @ingroup QkCircuit
+/// Return a list of string names for instructions in a circuit and their counts.
+///
+/// @param circuit A pointer to the circuit to get the counts for.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///     qk_circuit_count_ops(qc);
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> OpCounts {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    let count_ops = circuit.count_ops();
+    let mut output: Vec<OpCount> = count_ops
+        .into_iter()
+        .map(|(name, count)| OpCount {
+            name: CString::new(name).unwrap().into_raw(),
+            count,
+        })
+        .collect();
+    let data = output.as_mut_ptr();
+    let len = output.len();
+    std::mem::forget(output);
+    OpCounts { data, len }
+}
+
+/// @ingroup QkCircuit
+/// Return the number of instructions in the circuit
+///
+/// @param circuit A pointer to the circuit to get the counts for.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, QkGate_H, *[0], *[]);
+///     qk_circuit_num_instructions(qc); // 1
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_num_instructions(circuit: *const CircuitData) -> usize {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+    circuit.__len__()
+}
+
+#[repr(C)]
+pub struct CInstruction {
+    name: *mut c_char,
+    num_qubits: u32,
+    qubits: *mut u32,
+    num_clbits: u32,
+    clbits: *mut u32,
+    num_params: u32,
+    params: *mut f64,
+}
+
+/// @ingroup QkCircuit
+/// Return the number of instructions in the circuit
+///
+/// @param circuit A pointer to the circuit to get the counts for.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, QkGate_H, *[0], *[]);
+///     QkCircuitInstruction inst = qk_circuit_get_instruction(qc, 0);
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
```

Comment:
I realized this far into reading that this docstring is missing an "if"? There's also many instances of that so I'll just give you this one first 😄 
```suggestion
/// Behavior is undefined if ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
```

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/14006/commits/92a4a8f5d4cf08600b9d875534948f41fd693ce5